lldb/tools/lldb-rpc/lldb-rpc-gen/lldb-rpc-gen.cpp

```diff
@@ -0,0 +1,549 @@
+        CreateOutputFile(GetLibraryOutputDirectory(), LibraryHeaderFilename);
+    if (!LibraryHeaderOutputFile)
+      return nullptr;
+
+    ServerMethodOutputFile->keep();
+    ServerHeaderOutputFile->keep();
+    LibrarySourceOutputFile->keep();
+    LibraryHeaderOutputFile->keep();
+    return std::make_unique<SBConsumer>(
+        Byproducts, CI.getSourceManager(), CI.getASTContext(),
+        std::move(ServerMethodOutputFile), std::move(ServerHeaderOutputFile),
+        std::move(LibrarySourceOutputFile), std::move(LibraryHeaderOutputFile),
+        ClientCallbacksSourceEmitter, BindingsHarnessEmitter);
+  }
+
+private:
+  GeneratedByproducts &Byproducts;
+  lldb_rpc_gen::RPCBindingsHarnessEmitter &BindingsHarnessEmitter;
+  lldb_rpc_gen::RPCClientCallbacksSourceEmitter &ClientCallbacksSourceEmitter;
+};
+
+class SBActionFactory : public FrontendActionFactory {
+public:
+  SBActionFactory(
+      GeneratedByproducts &Byproducts,
+      lldb_rpc_gen::RPCBindingsHarnessEmitter &BindingsHarnessEmitter,
+      lldb_rpc_gen::RPCClientCallbacksSourceEmitter
+          &ClientCallbacksSourceEmitter)
+      : Byproducts(Byproducts), BindingsHarnessEmitter(BindingsHarnessEmitter),
+        ClientCallbacksSourceEmitter(ClientCallbacksSourceEmitter) {}
+
+  std::unique_ptr<FrontendAction> create() override {
+    return std::make_unique<SBAction>(Byproducts, BindingsHarnessEmitter,
+                                      ClientCallbacksSourceEmitter);
+  }
+
+private:
+  GeneratedByproducts &Byproducts;
+  lldb_rpc_gen::RPCBindingsHarnessEmitter &BindingsHarnessEmitter;
+  lldb_rpc_gen::RPCClientCallbacksSourceEmitter &ClientCallbacksSourceEmitter;
+};
+
+bool EmitAmalgamatedServerHeader(const std::vector<std::string> &Files) {
+  // Create the file
+  static constexpr llvm::StringLiteral AmalgamatedServerHeaderName = "SBAPI.h";
+  std::unique_ptr<llvm::ToolOutputFile> AmalgamatedServerHeader =
+      CreateOutputFile(GetServerOutputDirectory(), AmalgamatedServerHeaderName);
+  if (!AmalgamatedServerHeader)
+    return false;
+
+  // Write the header
+  AmalgamatedServerHeader->os()
+      << "#ifndef GENERATED_LLDB_RPC_SERVER_SBAPI_H\n";
+  AmalgamatedServerHeader->os()
+      << "#define GENERATED_LLDB_RPC_SERVER_SBAPI_H\n";
+  for (const auto &File : Files) {
+    llvm::StringRef FilenameNoExt =
+        llvm::sys::path::stem(llvm::sys::path::filename(File));
+    const std::string ServerHeaderFilename =
+        "Server_" + FilenameNoExt.str() + ".h";
+
+    AmalgamatedServerHeader->os()
+        << "#include \"" + ServerHeaderFilename + "\"\n";
+  }
+  AmalgamatedServerHeader->os() << "#include \"SBAPIExtensions.h\"\n";
+  AmalgamatedServerHeader->os()
+      << "#endif // GENERATED_LLDB_RPC_SERVER_SBAPI_H\n";
+  AmalgamatedServerHeader->keep();
+  return true;
+}
+
+bool EmitAmalgamatedLibraryHeader(const std::vector<std::string> &Files) {
+  static constexpr llvm::StringLiteral AmalgamatedLibraryHeaderName =
```

Comment:
Could be folded into the line below, and add the parameter name like `/*filename=*/"LLDBRPC.h"`.